larpandoracontent/LArHelpers/LArDiscreteProbabilityHelper.cc

```diff
@@ -0,0 +1,129 @@
+            LArDiscreteProbabilityHelper::MakeRandomisedSample(t1,randomNumberGenerator),
+            LArDiscreteProbabilityHelper::MakeRandomisedSample(t2,randomNumberGenerator)));
+
+        if ((rRandomised-rNominal) > std::numeric_limits<float>::epsilon())
+            nExtreme++;
+    }
+
+    return static_cast<float>(nExtreme)/static_cast<float>(nPermutations);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename T>
+float LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromStudentTDistribution(const T &t1, 
+    const T &t2, const size_t nIntegrationSteps)
+{
+    const float correlation(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(t1,t2));
+    const float dof(static_cast<float>(LArDiscreteProbabilityHelper::GetSize(t1)) - 2.f);
+    const float tTestStatistic(correlation*sqrt(dof)/(sqrt(1.f - correlation*correlation)));
+    const float tDistCoeff(std::tgamma(0.5f*(dof + 1.f))/std::tgamma(0.5f*dof)/(std::sqrt(dof*M_PI)));
+
+    const float upperLimit(15.f);
+    const float dx((upperLimit - tTestStatistic)/static_cast<float>(nIntegrationSteps));
+    float integral(tDistCoeff*std::pow(1.f + tTestStatistic*tTestStatistic/dof, -0.5f*(dof + 1.f)) + 
+            tDistCoeff*std::pow(1.f + upperLimit*upperLimit/dof, -0.5f*(dof + 1.f)));
+    for (unsigned int iStep = 1; iStep < nIntegrationSteps; ++iStep)
+        integral+=2.f*tDistCoeff*std::pow( 
+            1.f + (tTestStatistic + static_cast<float>(iStep)*dx)*(tTestStatistic + static_cast<float>(iStep)*dx)/dof, -0.5f*(dof + 1.f));
+    integral *= dx/2.f;
+
+    return integral;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename T>
+float LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(const T &t1, const T &t2)
+{
+    if (LArDiscreteProbabilityHelper::GetSize(t1) != (LArDiscreteProbabilityHelper::GetSize(t2)))
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    if (2 > LArDiscreteProbabilityHelper::GetSize(t1))
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    const float mean1(LArDiscreteProbabilityHelper::CalculateMean(t1));
+    const float mean2(LArDiscreteProbabilityHelper::CalculateMean(t2));
+
+    float variance1(0.f), variance2(0.f), covariance(0.f);
+
+    for (size_t iElement = 0; iElement < LArDiscreteProbabilityHelper::GetSize(t1); ++iElement)
+    {
+        const float diff1(LArDiscreteProbabilityHelper::GetElement(t1,iElement) - mean1);
+        const float diff2(LArDiscreteProbabilityHelper::GetElement(t2,iElement) - mean2);
+
+        variance1 += diff1*diff1;
+        variance2 += diff2*diff2;
+        covariance += diff1*diff2;
+    }
+
+    const float sqrtVars(std::sqrt(variance1*variance2));
```

Comment:
Looks like can't be taking sqrt of a negative number here. Of course, implementation above could change in future.

Reply:
Statistics would break if the variances ever went negative.  That being said, the logic could break down the line.
I've added in checks to the individual variances to see if they are less than epsilon.  Unfortunately, checking `var1*var2 < epsilon` does throw the exception because a lot of the variances are already small (multiplying two small numbers makes them doubly small!)

One edge case I am thinking about now, what if someone passes a vector/object where all of the numbers are the same?  The variance is identically 0 for such a vector and would cause the exception to be thrown.  Is that what we want?  I'm inclined to say that it is what we want as we probably shouldn't be passing vectors which contain the same repeated value....